File: src/main/java/ru/xpendence/topjavagraduation/repository/ReviewRepository.java
package ru.xpendence.topjavagraduation.repository;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import ru.xpendence.topjavagraduation.entity.Review;

import java.util.Optional;

public interface ReviewRepository extends JpaRepository<Review, Long> {

    @Query("SELECT r FROM Review r JOIN FETCH r.user JOIN FETCH r.restaurant WHERE r.id = :id")
    Optional<Review> findByIdWithDetails(@Param("id") Long id);

    @Query("""
            SELECT r FROM Review r
            JOIN FETCH r.user
            JOIN FETCH r.restaurant
            WHERE r.user.id = :userId AND r.restaurant.id = :restaurantId
            """)
    Optional<Review> findByUserIdAndRestaurantId(
            @Param("userId") Long userId,
            @Param("restaurantId") Long restaurantId
    );

    boolean existsByUserIdAndRestaurantId(Long userId, Long restaurantId);

    @Query(
            value = """
                    SELECT r FROM Review r
                    JOIN FETCH r.user
                    JOIN FETCH r.restaurant
                    WHERE r.restaurant.id = :restaurantId
                    """,
            countQuery = "SELECT COUNT(r) FROM Review r WHERE r.restaurant.id = :restaurantId"
    )
    Page<Review> findAllByRestaurantId(@Param("restaurantId") Long restaurantId, Pageable pageable);

    @Query("SELECT AVG(r.rating) FROM Review r WHERE r.restaurant.id = :restaurantId")
    Double findAverageRatingByRestaurantId(@Param("restaurantId") Long restaurantId);

    long countByRestaurantId(Long restaurantId);
}

File: src/main/kotlin/ru/xpendence/topjavagraduation/repository/ReviewRepository.kt
package ru.xpendence.topjavagraduation.repository

import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import ru.xpendence.topjavagraduation.entity.Review
import java.util.*

/**
 * Доступ к отзывам через JDBC.
 */
interface ReviewRepository {

    /**
     * Сохраняет отзыв: вставляет новую запись или обновляет существующую по [Review.id].
     *
     * @param review отзыв; для вставки — без id, с user.id и restaurant.id
     * @return сохранённый отзыв (с присвоенным id при вставке)
     */
    fun save(review: Review): Review

    /**
     * Возвращает отзыв по идентификатору с загруженными пользователем и рестораном.
     *
     * @param id идентификатор отзыва
     * @return отзыв с деталями или empty, если не найден
     */
    fun findByIdWithDetails(id: Long): Optional<Review>

    /**
     * Возвращает отзыв пользователя о ресторане с загруженными пользователем и рестораном.
     *
     * @param userId идентификатор пользователя
     * @param restaurantId идентификатор ресторана
     * @return отзыв или empty, если не найден
     */
    fun findByUserIdAndRestaurantId(userId: Long, restaurantId: Long): Optional<Review>

    /**
     * Проверяет наличие отзыва пользователя о ресторане.
     *
     * @param userId идентификатор пользователя
     * @param restaurantId идентификатор ресторана
     * @return true, если отзыв уже существует
     */
    fun existsByUserIdAndRestaurantId(userId: Long, restaurantId: Long): Boolean

    /**
     * Возвращает страницу отзывов ресторана с загруженными пользователем и рестораном.
     *
     * @param restaurantId идентификатор ресторана
     * @param pageable параметры страницы
     * @return страница отзывов
     */
    fun findAllByRestaurantId(restaurantId: Long, pageable: Pageable): Page<Review>

    /**
     * Возвращает средний рейтинг отзывов ресторана.
     *
     * @param restaurantId идентификатор ресторана
     * @return среднее значение или null, если отзывов нет
     */
    fun findAverageRatingByRestaurantId(restaurantId: Long): Double?

    /**
     * Возвращает число отзывов ресторана.
     *
     * @param restaurantId идентификатор ресторана
     * @return количество отзывов
     */
    fun countByRestaurantId(restaurantId: Long): Long

    /**
     * Проверяет существование отзыва по идентификатору.
     *
     * @param id идентификатор отзыва
     * @return true, если отзыв есть в БД
     */
    fun existsById(id: Long): Boolean

    /**
     * Удаляет отзыв по идентификатору.
     *
     * @param id идентификатор отзыва
     */
    fun deleteById(id: Long)
}
